### PR TITLE
Add notes for System.Binary

### PR DESCRIPTION
**Pull request by terrajobst**
@KrzysztofCwalina @stephentoub @karelz @ahsonkhan

@dotnet/fxdc

### CODE REVIEW

File: 2017/10-03-System.Binary/README.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+The family of `Span<T>` exists across these assemblies:
+
+1. `System.Memory`
+    - OOB `Span<T>`
+    - Later type forwarded to the core library where span is in-box
+2. `System.Buffers`
+    - Already contains the existing V1 `ArrayPool`
+
+(1) has to be in-box
+(2) doesn't have to be, but might due to `ArrayPool`. In fact, on .NET Core we
```
Comment (ahsonkhan) [inline 2017/10-03-System.Binary/README.md]: > In fact, on .NET Core we `ArrayPool` lives in `System.Private.Corelib`

reword

File: 2017/10-03-System.Binary/README.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+| [Issue](https://github.com/dotnet/corefx/issues/24144 )
+| [API Ref](System.Binary.md)
+| [Video](https://www.youtube.com/watch?v=m4BUM3nJZRw&t=404s)
+
+## Notes
+
+* Extension methods
+    - If we want to expose them as extension methods we need to have both
+      overloads for `ReadOnlySpan<T>` and `Span<T>`
+    - Let's do non-extension methods for now; we can make them extension methods
```
Comment (ahsonkhan) [inline 2017/10-03-System.Binary/README.md]: Does this apply to the {Try}Read{Xxx}/{Try}Write{Xxx} APIs only? Will the ReverseEndianness APIs remain as extension methods on the numeric primitive types?

Comment (terrajobst) [reply 2017/10-03-System.Binary/README.md]: My understanding ist that we decided that none of the APIs should be extension methods for now.

File: 2017/10-03-System.Binary/README.md (hunk 1 above)
Comment (terrajobst) [reply 2017/10-03-System.Binary/README.md]: Thanks, fixed

File: 2017/10-03-System.Binary/README.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+      later.
+        + But let's leave the type name to be `BufferPrimitives` (as opposed to
```
Comment (ahsonkhan) [inline 2017/10-03-System.Binary/README.md]: Should this be `BinaryPrimitives`?

Comment (terrajobst) [reply 2017/10-03-System.Binary/README.md]: Good catch! Yes.

File: 2017/10-03-System.Binary/README.md (hunk 2 above)
Comment (KrzysztofCwalina) [reply 2017/10-03-System.Binary/README.md]: Correct.

File: 2017/10-03-System.Binary/README.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    - `Span<T>`
+    - `Memory<T>`
+* `System.Buffers`
+    - `OwnedMemory<T>`
+    - Lifetime managers
+* `System.Buffers.Text`
+    - Formatting & parsing
+* `System.Buffers.Binary`
+    - The type discussed here
+    - `BinaryExtensions`
```
Comment (KrzysztofCwalina) [inline 2017/10-03-System.Binary/README.md]: BinaryPrimitives

File: 2017/10-03-System.Binary/README.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+* `Read<T>`/`Write<T>`
+    - Can we remove the specialized reader/writer methods in favor of the
+      generic `Read<T>`/`Write<T>` method?
+        + No, because the generic method cannot handle endianness.
+    - `Read<T>`/`Write<T>` should be renamed so that the name indicates it uses
+      machine endianness
+    - Should `Read<T>` return a `ref T`?
+        + No, because that wouldn't guarantee any alignment. Returning a `T`
+          guarantees that it's aligned because we always align the stack
+    - Should `Write<T>` take a `ref T`?
```
Comment (ahsonkhan) [inline 2017/10-03-System.Binary/README.md]: Should `Write<T>` take a `ref Span<byte>`?

Comment (ahsonkhan) [inline 2017/10-03-System.Binary/README.md]: Should `Write<T>` take a `ref Span<byte>`?

Comment (terrajobst) [reply 2017/10-03-System.Binary/README.md]: Why? You'd call it like this:

```C#
Write<MyStruct>(span, ref myStruct);
```

Passing the span by ref doesn't seem to buy us much, does it?

Comment (benaadams) [reply 2017/10-03-System.Binary/README.md]: Should it take an `in T`/`readonly ref T`?

File: 2017/10-03-System.Binary/README.md (hunk 4 above)
Comment (terrajobst) [reply 2017/10-03-System.Binary/README.md]: I think we decided to rename it to `BufferExtensions`?

File: 2017/10-03-System.Binary/README.md (hunk 5 above)
Comment (terrajobst) [reply 2017/10-03-System.Binary/README.md]: Good point. Will change that.

File: 2017/10-03-System.Binary/README.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+Thus, using `System.Buffers.Binary` feels very consistent.
+
+## Assemblies
+
+The family of `Span<T>` exists across these assemblies:
+
+1. `System.Memory`
+    - OOB `Span<T>`
+    - Later type forwarded to the core library where span is in-box
+2. `System.Buffers`
```
Comment (ahsonkhan) [inline 2017/10-03-System.Binary/README.md]: Should we update this given we are no longer adding them to System.Buffers.dll?

Comment (terrajobst) [reply 2017/10-03-System.Binary/README.md]: No, these are notes. There is no point in updating them when decisions are changing. That's why they are in directories that are tagged with a time.

**Pull request merged**